File: 0042-trapping-rain-water/solution.java
class Solution {
    public int trap(int[] height) {
        int total =0;
        int left=0, right = height.length-1;
        int mR=0, mL=0;
        while(left<right){
            mL = Math.max(mL,height[left]);
            mR = Math.max(mR,height[right]);
            if(mL < mR){
                total += mL - height[left];
                left++;
            }
            else{
                total += mR - height[right];
                right--;
            }
        }
        return total;
    }
}

File: 0042-trapping-rain-water/solution.py
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        rain = 0
        l , r = 0 , len(height) - 1
        mL = height [l]
        mR = height [r]

        while l < r:
            if mL < mR :
                l+=1
                mL = max(mL,height[l])
                rain += mL - height[l]
            else : 
                r-=1
                mR = max(mR,height[r])
                rain += mR - height[r]
        return rain
